enterprise/glassfish.common/src/org/netbeans/modules/glassfish/common/LogViewMgr.java

```diff
@@ -135,14 +147,11 @@ private LogViewMgr(final String uri) {
      */
     public static LogViewMgr getInstance(String uri) {
         LogViewMgr logViewMgr;
-        synchronized (instances) {
-            WeakReference<LogViewMgr> viewRef = instances.get(uri);
-            logViewMgr = viewRef != null ? viewRef.get() : null;
-            if(logViewMgr == null) {
-                logViewMgr = new LogViewMgr(uri);
-                instances.put(uri, new WeakReference<LogViewMgr>(logViewMgr));
-            }
-        }
+        
+        logViewMgr = instances.computeIfAbsent(uri, key -> 
+                        new WeakReference<LogViewMgr>(new LogViewMgr(uri))
+                    ).get();
```

Comment:
This is wrong. The value is a `WeakReference`. Line 140 in the old code fetch the referenced objects, effectively creating a new hardlink for it. If that fails (`get` returns `null`), then a new instance is created and a new `WeakReference` is created.
The supplier fed into `ConcurrentMap#computeIfAbsent` will only be invoked, if the `WeakReference` object is removed, which is not what was checked before.

Reply:
The supplier will be executed if the key is absent in the map.

I added some logging to this method and test both versions and the behavior is the same, this method triggers when starting, restarting, and stopping the server or when selecting the option `View Domain Server Log`:

![Screenshot from 2023-10-21 19-25-36](https://github.com/apache/netbeans/assets/9832133/ed7b6db7-b20f-43d7-8a1e-ebef966d46cd)

The first time it runs it will:
1. search the key(uri) in the map, because is the first time it runs with this GlassFish instance it will return `null`
2. compute the `value` to be inserted with the provided key (uri)
3. insert the `[key, value]` entry into the map
4. return the `value` which is a `WeakReference`
5. execute the `.get()` method from the returned `value`
6. assign the `LogViewMgr` obtained in step 5
7. return the `LogViewMgr` object

```
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: uri: [/path/to/app/server/glassfish625/glassfish:/path/to/app/server/glassfish625/glassfish/domains/domain1]deployer:gfv610ee9:localhost:4848
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: instances.get(uri): null
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: instances.get(uri).get(): null
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: instances.get(uri): java.lang.ref.WeakReference@766adea5
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: logViewMgr: org.netbeans.modules.glassfish.common.LogViewMgr@766adea3
```
The subsequent calls to the same GlassFish instance will log the same:
1. search the key(uri) in the map, it exists in the map
2. return the `value` associated to the key, a `WeakReference` object
3. execute the `.get()` method from the returned `value`
4. assign the `LogViewMgr` obtained in step 3
5. return the `LogViewMgr` object

```
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: uri: [/path/to/app/server/glassfish625/glassfish:/path/to/app/server/glassfish625/glassfish/domains/domain1]deployer:gfv610ee9:localhost:4848
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: instances.get(uri): java.lang.ref.WeakReference@766adea5
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: instances.get(uri).get(): org.netbeans.modules.glassfish.common.LogViewMgr@766adea3
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: instances.get(uri): java.lang.ref.WeakReference@766adea5
INFO [org.netbeans.modules.glassfish.common.LogViewMgr]: logViewMgr: org.netbeans.modules.glassfish.common.LogViewMgr@766adea3
```

I tested with two different GlassFish instances (version 6 and 7).

Reply:
Testing does not help when reasoning about possible synchronisation issues. They have the tendency not to manifest in tests and show in real life. The argument must be based on what is guaranteed by the language.

With this change the following can happen:

1. Call to `LogViewMgr#getInstance` with a fixed URI `uri1` happens. There is no key for `uri1` in the map `instances` and this the supplier is invoked. It creates a new `LogViewMgr`, wraps it into a `WeakReference`, places it in `instances` and returns it to the caller, which unwraps it using `WeakReference#get` and returns it.
2. The system does something with the `LogViewMgr` instance and at some point removes the last hard reference to it
3. Garbage collections happens
4. `LogViewMgr#getInstance` with a fixed URI `uri1` is called again. Now there is a `WeakReference` for the key `uri1`, which is returned and `WeakReference#get` is invoked on that. But because (3) happend you now get `null`. The code is broken.

There is another race in step 1. To better see this, the code can be split to:

```java
    public static LogViewMgr getInstance(String uri) {
        WeakRef<LogViewMgr> logViewMgrRef = instances.computeIfAbsent(uri, key -> 
                        new WeakReference<LogViewMgr>(new LogViewMgr(uri))
                    );
       LogViewMgr logViewMgr = logViewMgrRef.get();
        
        return logViewMgr;
    }
```

There is a very short window where the instance created in the supplier is not hard referenced anymore. After it is created and before `logViewMgrRef.get()` it is possible, that GC happens and the instance is collected. `logViewMgrRef.get()` will then yield `null`.

Reply:
Thank you for the explanation, the code example makes it easier to understand.

Reply:
Will this fix the hard reference issue?
```java
public static LogViewMgr getInstance(String uri) {
    LogViewMgr logViewMgr;
    logViewMgr = instances.computeIfAbsent(uri, (String key) -> {
                LogViewMgr _logViewMgr = new LogViewMgr(uri);
                WeakReference<LogViewMgr> _viewRef = new WeakReference<>(_logViewMgr);
                return _viewRef;
            }).get();
return logViewMgr;
```
Where is the race problem @matthiasblaesing? The `ConcurrentHashMap` implementation of `computeIfAbsent` is  atomic and concurrent.
[ConcurrentHashMap.computeIfAbsent](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent(K,java.util.function.Function))
> The entire method invocation is performed atomically. The supplied function is invoked exactly once per invocation of this method if the key is absent, else not at all. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple.

This will be executed in a `synchronized` block at a Node level instead of synchronizing the entire map.

Reply:
> Will this fix the hard reference issue?

No. That pretty much amounts to the same code.

> This will be executed in a synchronized block at a Node level instead of synchronizing the entire map.

Is there evidence that's a bottleneck?

Reply:
> Where is the race problem @matthiasblaesing? The `ConcurrentHashMap` implementation of `computeIfAbsent` is atomic and concurrent.

There are two races, so I don't know which of the two needs more explanation than what was provided.

Reply:
@neilcsmith-net I agree there is no bottleneck, just using a function that didn't fit the code pattern.

@matthiasblaesing After a long week, now I see the problem: the `.get()` is out of the safe `synchronized` block, my bad. Thank you both for the time and explanations.